### DISCUSSION TITLE
Using typeshed stubs

### DISCUSSION THREAD

**Discussion by ameyarao98** [microsoft/pyright#4805] [Q&A]
Hello, I am trying to use https://pypi.org/project/types-redis/ to type my project, but I am not entirely sure how :thinking: 
I use pdm, and have a virtual environment installed in the `.venv` of my project. I am wondering I what need to configure for these stubs to be used, I tried setting the stubPath but that did not dp anything

**Comment by erictraut** [comment]
The `types-redis` library is a "stub library" that you can install using pip. It is built and published from the source code in the typeshed repo. However, pyright already bundles all of the stubs in the typeshed repo, so you shouldn't need to install `types-redis`. You can simply `import redis` in your code, and pyright will use the bundled stubs.

**Reply by ameyarao98** [reply]
thanks for the quick response. This does indeed work when I import redis, but it doesn't when I use redis.asyncio for some reason

**Reply by erictraut** [reply]
According to the typeshed stubs, the top-level `redis` module does not import and then re-export the submodule `asyncio` from its namespace. To use the submodule, you'd need to import it explicitly, as in `import redis.asyncio`.

**Reply by ameyarao98** [reply]
that is what I currently do:
```
import redis.asyncio as redis

r = redis.from_url( 
    settings.redis_dsn, decode_responses=True
)
```
It doesn't recognise the stub.:
```
(function) def from_url(
    url: str,
    **kwargs: Unknown
) -> Redis[Any]
Returns an active Redis client generated from the given database URL.

Will attempt to extract the database id from the path url fragment, if none is provided.

Type of "from_url" is partially unknown
  Type of "from_url" is "(url: str, **kwargs: Unknown) -> Redis[Any]"Pylance[reportUnknownMemberType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnknownMemberType)
```

**Reply by erictraut** [reply]
It looks like you have strict-mode type checking enabled. The `reportUnknownMemberType` diagnostic rule isn't enabled by default in basic type checking. The reason you're receiving the error `"from_url" is partially unknown` is because the type stub is incomplete. It doesn't provide a type annotation for the `**kwargs` argument [here](https://github.com/python/typeshed/blob/8080e491d2fa92b1b1390c87fac23a1a38dcd919/stubs/redis/redis/asyncio/utils.pyi#L7). I recommend filing a bug (or even better, a PR) to the typeshed project to address the omission.

**Reply by ameyarao98** [reply]
ah indeed, will do! thanks for the help!
